phpBB/phpbb/report/factory.php

```diff
@@ -0,0 +1,88 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\report;
+
+class factory
+{
+	/**
+	 * @var \phpbb\db\driver\driver_interface
+	 */
+	protected $db;
+
+	/**
+	 * @var \phpbb\event\dispatcher_interface
+	 */
+	protected $dispatcher;
+
+	/**
+	 * @var \phpbb\config\db
+	 */
+	protected $config;
+
+	/**
+	 * @var \phpbb\auth\auth
+	 */
+	protected $auth;
+
+	/**
+	 * @var \phpbb\user
+	 */
+	protected $user;
+
+	/**
+	 * @var \phpbb\notification\manager
+	 */
+	protected $notifications;
+
+	/**
+	 * Constructor
+	 *
+	 * @param \phpbb\db\driver\driver_interface $db
+	 * @param \phpbb\event\dispatcher_interface $dispatcher
+	 * @param \phpbb\config\db $config
+	 * @param \phpbb\auth\auth $auth
+	 * @param \phpbb\user $user
+	 * @param \phpbb\notification\manager $notification
+	 */
+	public function __construct(\phpbb\db\driver\driver_interface $db, \phpbb\event\dispatcher_interface $dispatcher, \phpbb\config\db $config, \phpbb\auth\auth $auth, \phpbb\user $user, \phpbb\notification\manager $notification)
+	{
+		$this->db				= $db;
+		$this->dispatcher		= $dispatcher;
+		$this->config			= $config;
+		$this->auth				= $auth;
+		$this->user				= $user;
+		$this->notifications	= $notification;
+	}
+
+	/**
+	 * Report factory
+	 *
+	 * @param string	$type
+	 * @return \phpbb\report\report_handler_interface
+	 */
+	public function get($type)
+	{
+		switch ($type)
+		{
+			case 'pm':
+				return new report_handler_pm($this->db, $this->dispatcher, $this->config, $this->auth, $this->user, $this->notifications);
+				break;
+			case 'post':
+				return new report_handler_post($this->db, $this->dispatcher, $this->config, $this->auth, $this->user, $this->notifications);
+				break;
+		}
+
+		throw new \InvalidArgumentException('Invalid database driver given');
```

Comment:
Use a language string and a sub class of \phpbb\exception\runtime_exception (you an create the class under the same namespace)